tests/cases/auth_context.rs

```diff
@@ -0,0 +1,293 @@
+        .register_service(service)
+        .bind_secure("127.0.0.1", 0, server_credentials)
+        .build()
+        .unwrap();
+    server.start();
+    let port = server.bind_addrs()[0].1;
+
+    let client_credentials = ChannelCredentialsBuilder::new()
+        .root_cert(CA_CRT.as_bytes().to_owned())
+        .cert(
+            CLIENT_CRT.as_bytes().to_owned(),
+            CLIENT_KEY.as_bytes().to_owned(),
+        )
+        .build();
+    let ch =
+        ChannelBuilder::new(env).secure_connect(&format!("localhost:{}", port), client_credentials);
+    let client = GreeterClient::new(ch);
+
+    let mut req = HelloRequest::new_();
+    req.set_name("world".to_owned());
+    let resp = client.say_hello(&req).unwrap();
+
+    assert_eq!(resp.get_message(), "hello world");
+
+    let keys = rx.recv_timeout(Duration::from_secs(1)).unwrap();
+    assert_eq!(keys, ("AuthContextPresent".to_owned(), "Y".to_owned()));
+    // Test auth_context keys
+    let keys = rx.recv_timeout(Duration::from_secs(1)).unwrap();
+    assert_eq!(
+        keys,
+        ("transport_security_type".to_owned(), "ssl".to_owned())
+    );
+    let keys = rx.recv_timeout(Duration::from_secs(1)).unwrap();
+    assert_eq!(
+        keys,
+        (
+            "x509_common_name".to_owned(),
+            "grpc-rs Test Client".to_owned()
+        )
+    );
+    let keys = rx.recv_timeout(Duration::from_secs(1)).unwrap();
+    assert_eq!(
+        keys,
+        ("x509_pem_cert".to_owned(), CLIENT_CRT.to_owned() + "\n")
+    );
+    let keys = rx.recv_timeout(Duration::from_secs(1)).unwrap();
+    assert_eq!(keys, ("ssl_session_reused".to_owned(), "false".to_owned()));
+    let _empty_keys: mpsc::RecvTimeoutError = rx
+        .recv_timeout(Duration::from_millis(100))
+        .expect_err("Received more auth_context vars than expected");
+}
+
+#[test]
+fn test_no_crash_on_insecure() {
+    let env = Arc::new(EnvBuilder::new().build());
+    let (tx, rx) = mpsc::channel();
+    let service = create_greeter(GreeterService { tx: tx });
+    let mut server = ServerBuilder::new(env.clone())
+        .register_service(service)
+        .bind("127.0.0.1", 0)
+        .build()
+        .unwrap();
+    server.start();
+    let port = server.bind_addrs()[0].1;
+
+    let ch = ChannelBuilder::new(env).connect(&format!("localhost:{}", port));
+    let client = GreeterClient::new(ch);
+
+    let mut req = HelloRequest::new_();
+    req.set_name("world".to_owned());
+    let resp = client.say_hello(&req).unwrap();
+
+    assert_eq!(resp.get_message(), "hello world");
+
+    // Test auth_context keys
+    let keys = rx.recv_timeout(Duration::from_secs(1)).unwrap();
+    assert_eq!(keys, ("AuthContextPresent".to_owned(), "N".to_owned()));
+    let _empty_keys: mpsc::RecvTimeoutError = rx
+        .recv_timeout(Duration::from_millis(100))
+        .expect_err("Received auth context even though not authenticated");
+}
+
+static CA_CRT: &str = r#"-----BEGIN CERTIFICATE-----
```

Comment:
Can you reuse the certificates already exists in this repository?

Reply:
Maybe, I don't know, I didn't spend time looking for any and analyzing whether I could use them, whether they would last forever, whether the names that we now have access to were set in a meaningful way... I was however pretty sure of the amount of time it would take me to generate a set of them with exactly the characteristics I needed.

Reply:
The test case will fail if the file is missing. Reusing the certificates can make the sources small and readable. You can just use `grpc_proto::util::create_test_channel_credentials` and `grpc_proto::util::create_test_server_credentials` to create the credentials you need.

Reply:
I need a CA and two cert/key pairs, one of them for my client, hence the existing key pairs are not enough for that test.
Also, I need the credentials crated to have both the CA and the cert/key (since both authenticate each other, the goal being to authenticate the client), hence these functions do not fit my need.
I find it convenient to just define them in the file and I don't feel like it hurts its readability but if you want to refactor that part, feel free to do so (I took great care on the code but I don't have enough time atm to just spend it all on writing the most perfect tests ever.)

Reply:
Just one CA is enough. What need's to do is add another pair of cert/key.